{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from read_gff import read_gff\n",
    "data_dir = '../data'\n",
    "ref_dir = f'{data_dir}/reference'\n",
    "pgt_201_b1_gff_path = f'{ref_dir}/Pgt_201_B1_GeneCatalog_20191015.gff3'\n",
    "gff = read_gff(pgt_201_b1_gff_path)\n",
    "hit_to_gene = {\n",
    "    'jgi.p|Pgt_201_B1|9243': 'SdhA',\n",
    "    'jgi.p|Pgt_201_B1|44': 'SdhB',\n",
    "    'jgi.p|Pgt_201_B1|1581': 'SdhC',\n",
    "    'jgi.p|Pgt_201_B1|9290': 'SdhD',\n",
    "    'jgi.p|Pgt_201_B1|17472': 'Cyp51',\n",
    "}\n",
    "hits_gff = gff[gff.Name.isin(hit_to_gene) & (gff['type'] == 'gene')]\n",
    "hits_descendants = set(s for l in hits_gff.Descendants.values for s in l)\n",
    "hits_gff = gff[gff.Name.isin(hit_to_gene) | gff.ID.isin(hits_descendants)]\n",
    "gff = pd.read_csv(pgt_201_b1_gff_path, sep='\\t', header=None, comment='#')\n",
    "hits_gff = gff.loc[hits_gff.index]\n",
    "hits_gff.to_csv(f'{ref_dir}/fungicide_relative_to_pgt_201_b1.gff3', sep='\\t', index=None, header=None)\n",
    "\n",
    "hits_gff.columns = ['seqid', 'source', 'type', 'start', 'end', 'score', 'strand', 'phase', 'attributes']\n",
    "import re\n",
    "from Bio.Seq import reverse_complement\n",
    "from Bio.SeqIO import parse\n",
    "ref = {r.id: str(r.seq) for r in parse(f'{ref_dir}/Pgt_201_B1_AssemblyScaffolds2.fasta', 'fasta')}\n",
    "\n",
    "# write the fungicide gene sequences as fasta\n",
    "with open(f'{ref_dir}/fungicide_genes.fa', 'w') as f:\n",
    "    for _, row in hits_gff.query('type == \"gene\"').iterrows():\n",
    "        hit_name =  next(re.finditer('Name=([^;]+);', row.attributes)).group(1)\n",
    "        gene = hit_to_gene[hit_name]\n",
    "        if row.strand == '-':\n",
    "            gene_seq = reverse_complement(ref[row.seqid][row.start - 1:row.end])\n",
    "        else:\n",
    "            gene_seq = ref[row.seqid][row.start - 1:row.end]\n",
    "        f.write('>' + gene + '\\n' + gene_seq + '\\n')\n",
    "\n",
    "hit_names = []\n",
    "for _, row in hits_gff.iterrows():\n",
    "    if row['type'] == 'gene':\n",
    "        hit_name = row['attributes'].split('Name=')[1].split(';')[0]\n",
    "    hit_names.append(hit_name)\n",
    "genes = [hit_to_gene[hit_name] for hit_name in hit_names]\n",
    "hits_gff['seqid'] = genes\n",
    "\n",
    "genes_gff = []\n",
    "for gene, gene_gff in hits_gff.groupby('seqid'):\n",
    "    gene_gff = gene_gff[gene_gff['type'].isin({'gene', 'mRNA', 'exon', 'CDS'})].copy()\n",
    "    # Make relative to gene not chromosome\n",
    "    old_start = gene_gff.start.values[0]\n",
    "    old_end = gene_gff.end.values[0]\n",
    "    length = old_end - old_start + 1\n",
    "    old_strand = gene_gff.strand.values[0]\n",
    "    gene_gff['start'] = gene_gff.start + 1 - old_start\n",
    "    gene_gff['end'] = gene_gff.end + 1 - old_start\n",
    "    gene_gff_non_mrna_descendants = gene_gff[~ gene_gff['type'].isin({'exon', 'CDS'})].copy()\n",
    "    gene_gff_mrna_descendants = gene_gff[gene_gff['type'].isin({'exon', 'CDS'})].copy()\n",
    "    if old_strand == '-':\n",
    "        # display(gene_gff_mrna_descendants)\n",
    "        new_start = length - gene_gff_mrna_descendants.end + 1\n",
    "        new_end = length - gene_gff_mrna_descendants.start + 1\n",
    "        gene_gff_mrna_descendants['start'] = new_start\n",
    "        gene_gff_mrna_descendants['end'] = new_end\n",
    "        # display(gene_gff_mrna_descendants)\n",
    "    gene_gff = pd.concat([gene_gff_non_mrna_descendants, gene_gff_mrna_descendants])\n",
    "    gene_gff['strand'] = '+'\n",
    "\n",
    "    phases = []\n",
    "    n = 0\n",
    "    for _, row in gene_gff.iterrows():\n",
    "        if row.type != 'CDS':\n",
    "            phases.append('.')\n",
    "        else:\n",
    "            phases.append(2 - (n - 1) % 3)\n",
    "            n += 1 + (row.end - row.start)\n",
    "    gene_gff['phase'] = phases\n",
    "    genes_gff.append(gene_gff)\n",
    "\n",
    "genes_gff = pd.concat(genes_gff)\n",
    "genes_gff.to_csv('../data/reference/fungicide_genes.gff3', sep='\\t', header=None, index=None)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.8.5 ('bio')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.13"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "cb121900eef4c2628aef9ca2784e6d2a1b4be0c89e4ac151916a8e0edb5a77f5"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
